go/vt/vtgate/evalengine/arithmetic_expr.go

```diff
@@ -30,7 +30,7 @@ type (
 
 	// ArithmeticOp allows arithmetic expressions to not have to evaluate child expressions - this is done by the BinaryExpr
 	ArithmeticOp interface {
-		eval(left, right EvalResult) (EvalResult, error)
+		eval(left, right, out *EvalResult) error
```

Comment:
I don't understand this change. Why is the output of the function one of the input params?

Reply:
All the evaluation has been changed to take an `EvalResult` as an output argument instead of returning it because it opens up more possibilities for optimization. For simple evaluation cases, it allows us to pass the output result directly from our stack (assuming the compiler doesn't make it escape, which it does sometimes). For more complex cases, it'll allow us to use a memory pool for reusing temporary/intermediate evaluation results, something that couldn't be done if the result is returned directly by the evaluator.